src/runtime/formal/ops/ops.cc

```diff
@@ -42,27 +39,28 @@ CVM_REGISTER_GLOBAL("cvm.runtime.formal.dense")
   auto X_data = CVMArg2Data<int32_t>(X); 
   auto W_data = CVMArg2Data<int32_t>(W); 
   auto Y_data = CVMArg2Data<int32_t>(Y); 
-  // Y = X * WT
+  // Y = X * W^T
+  Indices xIdx(X_shape), wIdx(W_shape), yIdx(Y_shape);
   for (int64_t m = 0; m < Y_shape[0]; ++m) {
     // Y(m, n) = X(m, k) * WT(k, n) = X(m, k) * W(n, k)
-    int32_t Y_offset = m * Y_shape[1], X_offset = m * X_shape[1];
+    xIdx.Ref(0) = m;
     for (int64_t n = 0; n < Y_shape[1]; ++n) {
-      int32_t sum = 0, W_offset = n * W_shape[1];
+      wIdx.Ref(0) = n;
+      int32_t sum = 0;  //, W_offset = n * W_shape[1];
       for (int64_t k = 0; k < X_shape[1]; ++k) {
-        sum += X_data[X_offset + k] * W_data[W_offset + k];
+        xIdx.Ref(1) = k, wIdx.Ref(1) = k;
+        sum += X_data[xIdx.Index()] * W_data[wIdx.Index()];
       }
-      Y_data[Y_offset + n] = sum;
+      Y_data[yIdx.Index()] = sum;
+      yIdx++;
```

Comment:
Suggest that use the loop for `yIdx` directly, and get the `m` and `n` as the
val of `yIdx[0]` and `yIdx[1]` respectively.